File: 200-NumberOfIslands/200-NumberOfIslands.java
// Last updated: 6/10/2026, 7:25:51 PM
class Solution {
    public int numIslands(char[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        boolean[][] vis = new boolean[n][m];
        int[] dx = {1 , -1 , 0 , 0};
        int[] dy = {0 , 0 , 1 , -1};
        
        BiPredicate<Integer, Integer> check =
            (a, b) -> 0 <= a && a < n &&
                    0 <= b && b < m &&
                    grid[a][b] == '1' &&
                    !vis[a][b];
        int cnt = 0;
        for (int i=0 ; i<n ; i++) {
            for (int j=0 ; j<m ; j++) {
                if (grid[i][j] == '0' || vis[i][j]) continue;
                cnt++;
                if (!vis[i][j]) {
                    Queue<int[]> q = new LinkedList<>();
                    q.add(new int[] {i , j});
                    vis[i][j] = true;
                    while (!q.isEmpty()) {
                        var pos = q.poll();
                        int x = pos[0] , y = pos[1];
                        for (int d=0 ; d<4 ; d++) {
                            int nx = x + dx[d] , ny = y + dy[d];
                            if (check.test(nx , ny)) {
                                vis[nx][ny] = true;
                                q.add(new int[] {nx , ny});
                            }
                        }
                    }
                } 
            }
        }
        return cnt;
    }
}

File: 200-NumberOfIslands/200-NumberOfIslands.cpp
// Last updated: 6/10/2026, 7:25:57 PM
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size() , m = grid[0].size();
        function<void(int , int)> bfs = [&](int x , int y) {
            queue<pair<int , int>> q;
            q.push({x , y});
            grid[x][y] = '0';
            int dx[] = {1 , -1 , 0 , 0};
            int dy[] = {0 , 0 , 1 , -1};
            while(!q.empty()) {
                auto [x , y] = q.front(); q.pop();
                cout << x << ' ' << y << "\n";
                for(int i=0 ; i<4 ; i++) {
                    int nx = x + dx[i] , ny = y + dy[i];
                    if(nx >= 0 && nx < n && ny >= 0 && ny < m && grid[nx][ny] == '1') {
                        q.push({nx , ny});
                        grid[nx][ny] = '0';
                    }
                } 
            }   
        };
        int res = 0;
        // bfs(0 , 0);
        // for(int i=0 ; i<n ; i++) {
        //     for(int j=0 ; j<m ; j++)
        //         cout << grid[i][j] << ' ';
        //     cout << '\n';
        // }
        for(int i=0 ; i<n; i++) {
            for(int j=0 ; j<m ; j++) {
                if(grid[i][j] == '1') {
                    // cout << i << ' ' << j << "\n";
                    ++res;
                    bfs(i , j);
                }
            }
        }
        return res;
    }
};